File: src/types/type.ts
export type Type =
  | PrimitiveType
  | ArrayType
  | SequenceType
  | OptionalType
  | ErrorType
  | PointerType
  | NamedType
  | RecordType
  | ChoiceType
  | FunctionType
  | LiteralType
  | UnknownType;

export interface PrimitiveType {
  kind: "primitive";
  name: string;
  width?: number;
}

export interface ArrayType {
  kind: "array";
  elem: Type;
}

export interface SequenceType {
  kind: "sequence";
  elem: Type;
}

export interface OptionalType {
  kind: "optional";
  elem: Type;
}

export interface ErrorType {
  kind: "error";
  elem: Type;
}

export interface PointerType {
  kind: "pointer";
  elem: Type;
}

export interface NamedType {
  kind: "named";
  name: string;
}

export interface RecordType {
  kind: "record";
  name: string;
  fields: Map<string, Type>;
}

export interface ChoiceType {
  kind: "choice";
  name: string;
  variants: Map<string, Type | undefined>;
}

export interface FunctionType {
  kind: "function";
  params: Type[];
  ret: Type;
}

export interface LiteralType {
  kind: "literal";
  value: string;
}

export interface UnknownType {
  kind: "unknown";
}

export function typeToString(t: Type): string {
  switch (t.kind) {
    case "primitive":
      return t.width ? `${t.name} ${t.width}` : t.name;
    case "array":
      return `array of ${typeToString(t.elem)}`;
    case "sequence":
      return `sequence of ${typeToString(t.elem)}`;
    case "optional":
      return `optional ${typeToString(t.elem)}`;
    case "error":
      return `error ${typeToString(t.elem)}`;
    case "pointer":
      return `pointer to ${typeToString(t.elem)}`;
    case "named":
      return t.name;
    case "record":
      return t.name;
    case "choice":
      return t.name;
    case "function":
      const params = t.params.map(typeToString).join(", ");
      return `function(${params}) -> ${typeToString(t.ret)}`;
    case "literal":
      return `literal ${t.value}`;
    case "unknown":
      return "unknown";
  }
}

export function typesEqual(a: Type, b: Type): boolean {
  if (a.kind !== b.kind) return false;
  
  switch (a.kind) {
    case "primitive":
      return a.name === (b as PrimitiveType).name && a.width === (b as PrimitiveType).width;
    case "array":
      return typesEqual(a.elem, (b as ArrayType).elem);
    case "sequence":
      return typesEqual(a.elem, (b as SequenceType).elem);
    case "optional":
      return typesEqual(a.elem, (b as OptionalType).elem);
    case "error":
      return typesEqual(a.elem, (b as ErrorType).elem);
    case "pointer":
      return typesEqual(a.elem, (b as PointerType).elem);
    case "named":
      return a.name === (b as NamedType).name;
    case "record":
    case "choice":
      return a.name === (b as RecordType | ChoiceType).name;
    case "function":
      const bf = b as FunctionType;
      if (a.params.length !== bf.params.length) return false;
      for (let i = 0; i < a.params.length; i++) {
        if (!typesEqual(a.params[i], bf.params[i])) return false;
      }
      return typesEqual(a.ret, bf.ret);
    case "literal":
      return a.value === (b as LiteralType).value;
    case "unknown":
      return true;
  }
}

File: src/types/type.nim
import std/tables
import std/strutils
import ../sema/bindings

type
  Type* = ref object
    kind*: string
    name*: string
    width*: int
    elem*: Type
    constant*: bool
    fields*: OrderedTable[string, Type]
    variants*: OrderedTable[string, Type]
    params*: seq[Type]
    ret*: Type
    abi*: string
    arguments*: seq[Type]
    length*: int
    layout*: string
    value*: string
    borrows*: seq[int]
    generics*: seq[string]
    constraints*: seq[tuple[subject: string, trait: string]]

proc typeToString*(value: Type): string =
  if value == nil: return "unknown"
  case value.kind
  of "primitive": return if value.width > 0: value.name & " " & $value.width else: value.name
  of "array": return "array of " & typeToString(value.elem)
  of "sequence": return "sequence of " & (if value.constant: "constant " else: "") & typeToString(value.elem)
  of "optional": return "optional " & typeToString(value.elem)
  of "error": return "fallible " & typeToString(value.elem)
  of "pointer": return "pointer to " & typeToString(value.elem)
  of "named", "record", "choice", "union", "opaque": return value.name
  of "vector": return "vector[" & $value.length & ", " & typeToString(value.elem) & "]"
  of "function":
    var params: seq[string]
    for parameter in value.params: params.add(typeToString(parameter))
    return "function(" & params.join(", ") & ") -> " & typeToString(value.ret) & (if value.abi.len > 0: " for " & provider(value.abi) else: "")
  of "literal": return "literal " & value.value
  else: return "unknown"

proc typesEqual*(left, right: Type): bool =
  if left == nil or right == nil: return left == right
  if left.kind != right.kind: return false
  case left.kind
  of "primitive": left.name == right.name and (if left.width > 0: left.width else: 64) == (if right.width > 0: right.width else: 64)
  of "array", "optional", "error", "pointer": typesEqual(left.elem, right.elem)
  of "sequence": left.constant == right.constant and typesEqual(left.elem, right.elem)
  of "named", "record", "choice", "union", "opaque": left.name == right.name
  of "vector": left.length == right.length and typesEqual(left.elem, right.elem)
  of "function":
    if left.params.len != right.params.len or left.abi != right.abi or not typesEqual(left.ret, right.ret): false
    else:
      for index in 0 ..< left.params.len:
        if not typesEqual(left.params[index], right.params[index]): return false
      true
  of "literal": left.value == right.value
  of "unknown": true
  else: false
